Primary keys (#4986)
confluentinc/ksql: docs-md/developer-guide/ksqldb-reference/create-table.md

--------

```sql
CREATE TABLE table_name ( { column_name data_type (PRIMARY KEY) } [, ...] )
  WITH ( property_name = expression [, ...] );
```

Description
-----------

Create a new table with the specified columns and properties.

A ksqlDB TABLE works much like tables in other SQL systems. A table has zero or more rows. Each
row is identified by its `PRIMARY KEY`. `PRIMARY KEY` values can not be NULL. A message in the
underlying Kafka topic with the same key as an existing row will _replace_ the earlier row in the
table, or _delete_ the row if the message's value is NULL, as long as the earlier row does not have

> a later timestamp / `ROWTIME`. The table below contrasts this to a [ksqlDB STREAM](./create-stream):

Reviewer: ```suggestion
a later timestamp / `ROWTIME`. This situation is handled differently by a [ksqlDB STREAM](./create-stream), as shown in the following table.

```

|                          |  STREAM                                                       | TABLE                                                             |
| ------------------------ | --------------------------------------------------------------| ----------------------------------------------------------------- |
| Key column type          | `KEY`                                                         | `PRIMARY KEY`                                                     |

> | NON NULL key constraint  | No                                                            | Yes <br> Messages in the Kafka topic with a NULL `PRIMARY KEY` are ignored |

Reviewer: ```suggestion
| NON NULL key constraint  | No                                                            | Yes <br> Messages in the Kafka topic with a NULL `PRIMARY KEY` are ignored.|

```

> | Unique key constraint    | No <br> Messages with the same key as another have no special meaning | Yes <br> Later messages with the same key _replace_ earlier |

Reviewer: ```suggestion
| Unique key constraint    | No <br> Messages with the same key as another have no special meaning.| Yes <br> Later messages with the same key _replace_ earlier.|

```

> | Tombstones               | No <br> Messages with NULL values are ignored                 | Yes <br> NULL message values are treated as a _tombstone_ <br> Any existing row with a matching key is deleted |

Reviewer: ```suggestion
| Tombstones               | No <br> Messages with NULL values are ignored.                | Yes <br> NULL message values are treated as a _tombstone_. <br> Any existing row with a matching key is deleted. |

```

Each column is defined by:

>  * `column_name`: the name of the column. If unquoted the name must be a valid

Reviewer: ```suggestion
 * `column_name`: the name of the column. If unquoted, the name must be a valid

```

   [SQL identifier](../../concepts/schemas#valid-identifiers) and will be converted to uppercase.
   The name can be quoted if case needs to be preserved or if the name is not a valid SQL

>    identifier, for example ``` `mixedCaseId` ``` or ``` `$with@invalid!chars`.

Reviewer: ```suggestion
   identifier, for example ``` `mixedCaseId` ``` or ``` `$with@invalid!chars` ```.

```

 * `data_type`: the SQL type of the column. Columns can be any of the
   [data types](../syntax-reference.md#ksqldb-data-types) supported by ksqlDB.
 * `PRIMARY KEY`: columns that are stored in the Kafka message's key should be marked as

>    `PRIMARY KEY` columns. If a column is not marked as a `PRIMARY KEY` column ksqlDB will load it

Reviewer: ```suggestion
   `PRIMARY KEY` columns. If a column is not marked as a `PRIMARY KEY` column, ksqlDB loads it

```
